src/core/polyhedral/codegen_llvm.cc

```diff
@@ -351,6 +355,45 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
           << "[LLVM-IR] After optimization:\n"
           << toString(module.get());
     }
+
+    // In a separate process call the LLC and dump the ASM
+    // Since this is for debugging purposes only atm, separate process is fine.
+    if (FLAGS_llvm_dump_asm) {
+      // Create a temporary file (do not use racy tmpname;
+      // do not use tmpnam that immediately destroys the file).
+      char fileCName[256] = "/tmp/XXXXXXXXXX";
+      auto fd = mkstemp(fileCName);
+      ScopeGuard sg([fileCName]() { unlink(fileCName); });
+      CHECK_GE(fd, 0) << "mkstemp(): " << fileCName << " returned fd: " << fd;
+      close(fd);
+
+      // Close and reopen as FILE*
+      auto tmp = fopen(fileCName, "wb+");
+      auto llvmIr = toString(module.get());
+      std::fputs(llvmIr.c_str(), tmp);
+      fclose(tmp); // Close properly so the piped command can open without race
+
+      std::string fileName(fileCName);
+      // TODO: evolve this in a better way to find LLVM
+      auto command = std::string(" `which llc || which llc-5.0` ") +
+          std::string(" -asm-verbose ") + std::string(" < ") + fileName;
+      FILE* fpipe = popen(command.c_str(), "r");
+      if (fpipe == nullptr) {
+        CHECK(false) << "Failed to popen() on command: " << command;
+      }
+      ScopeGuard sg2([fpipe]() { pclose(fpipe); });
```

Comment:
Hmm... I would just close it after I finished reading from it.

Reply:
This does almost exactly that, via scoping / RAII (ok technically```LOG(INFO)``` is actually printed before the close). I tend to prefer to keep related alloc/free close in code and "be done with them" early. This is also generally robust to control-flow, even though it is not a factor here.